text/0000-metabuild.mdwn

```diff
@@ -0,0 +1,131 @@
+`build.rs` file. Support experimentation with declarative build scripts in the
+crates.io ecosystem.
+
+# Motivation
+
+Cargo has many potentially desirable enhancements planned for its build
+process, including integrating a Cargo build process with native dependencies,
+and integrating with broader build systems or projects, such as massive
+mono-repo build systems, or Linux distributions.
+
+Right now, the biggest problem facing such systems involves `build.rs` scripts
+and the arbitrary things those scripts can do. Such build systems typically
+need more information about native dependencies that are embedded in
+`build.rs`, so that they can provide their own versions of those dependencies,
+or encode appropriate dependencies in another metadata format such as the
+dependencies of their packaging system or build system. Right now, such systems
+often have to override the `build.rs` script themselves, and do custom
+per-crate integration work, manually; there’s no way to introspect what
+`build.rs` does, or get a declarative semantic description of the build script.
+
+At the same time, we don't yet have sufficiently precise information about the
+needs of such systems to design an ideal set of Cargo metadata on the first
+try. Rather than attempt to architect the perfect solution from the start, and
+potentially create an intermediate state that will require long-term support,
+we propose to allow experimentation with declarative build systems within the
+crates.io ecosystem, in crates supplying modular components similar to
+`build.rs` scripts. By convention, such scripts should typically read any
+parameters and metadata they need from `Cargo.toml`, in a form that other
+build-related software can read as well.
+
+# Guide-level explanation
+
+In the `[package]` section of `Cargo.toml`, you can specify a field
+`metabuild`, whose value should be a string or list of strings, each one
+exactly matching the name of a dependency specified in the
+`[build-dependencies]` section. If you specify `metabuild`, you must not
+specify `build`, and Cargo will ignore the `build.rs` file if any.
+
+When Cargo builds a crate that specifies a `metabuild` field, at the point when
+it would have built and run `build.rs`, it will instead invoke the
+`metabuild()` function from each of the specified crates in order.
+
+In effect, Cargo will act as though it had a `build.rs` file containing an
+`extern crate` line for each string, in order, as well as a `main` function
+that calls the `metabuild` function in each such crate, in order. For example,
+if the crate contains `metabuild = ["pkgc", "parsegen"]`, then the effective
+`build.rs` will look like this:
+
+```rust
+extern crate pkgc;
+extern crate parsegen;
+
+fn main() {
+    pkgc::metabuild();
+    parsegen::metabuild();
+}
+```
+
+Note that the `metabuild` functions intentionally take no parameters; they
+should obtain any parameters they need from `Cargo.toml`. Various crates to
```

Comment:
I assume this refers to http://doc.crates.io/manifest.html#the-metadata-table-optional ? It’d be nice for the RFC to provide an example `fn metabuild()` implementation that does so.